### PR TITLE
rng-tools: start as early as possible

### PR DESCRIPTION
**Pull request by Noltari**
Maintainer: @hnyman @nwf @pprindeville 
Compile tested: bcm27xx/bcm2708
Run tested: bcm27xx/bcm2708 (Raspberry Pi B+)

Description:
Raspberry Pi devices (bcm2708) need this to get from:
[  102.310494] random: crng init done
to:
[   12.539744] random: crng init done

### CODE REVIEW

**Comment by neheb** [comment]
How about haveged?

**Comment by pprindeville** [comment]
> How about haveged?

Eh?

File: utils/rng-tools/files/rngd.init [hunk 1]
```diff
@@ -1,7 +1,7 @@
 #!/bin/sh /etc/rc.common
 # Copyright (C) 2011-2014 OpenWrt.org
 
-START=25
+START=0
```
Comment (pprindeville) [inline utils/rng-tools/files/rngd.init:4]: Zero seems a bit early.  I don't like the idea of not being able to put something before this...  What about `5`?

**Comment by Noltari** [comment]
> How about haveged?

@neheb haveged works because it has START=01:
https://github.com/openwrt/packages/blob/4b6948bbf01989c2c299a64748485b7b65edcbf2/utils/haveged/files/haveged.init#L3

@pprindeville do you think START=1 is OK?
BTW, urngd is using START=0

**Comment by pprindeville** [comment]
> BTW, urngd is using START=0

Okay, let's give that a try then.

**Review by pprindeville — APPROVED** [review]

**Review by neheb — APPROVED** [review]

**Comment by hnyman** [comment]
> BTW, urngd is using START=0

Actually, it is using 00

All priorities so far have been done with two digits (00-99). I think that rng-tools should be so, too.
The run order is alphabetical, so it would now be:  S00sysfixtime, S00urngd, S0rng-tools

I also think that all relevant code, e.g. in rc.common expects a two-digit priority:

https://github.com/openwrt/openwrt/blob/3f1109bf2af2cf4977689a33ea481afaaac8dc98/package/base-files/files/etc/rc.common#L48

`		ln -sf "../init.d/$name" "$IPKG_INSTROOT/etc/rc.d/S${START}${name##S[0-9][0-9]}" && \`

**Comment by hnyman** [comment]
> merged commit f5a615b

Hmm. Just missed with my comment :-(

**Comment by Noltari** [comment]
> > merged commit [f5a615b](https://github.com/openwrt/packages/commit/f5a615b32c67646df21e14686912c1650d4860a1)
> 
> Hmm. Just missed with my comment :-(

Ah shit, you're right...
Should I open a new PR to fix this?

**Comment by hnyman** [comment]
Just commit a fix change from 0 to 00.
No need to have a PR

**Comment by tofurky** [comment]
starting rngd before kmods are loaded prevents the use of any rng that requires a kmod to be loaded before its device appears, e.g. ttyACM0.

moving it to START=11 allows it to be used with devices like this. would that be a good compromise between starting early enough and remaining compatible for hwrng drivers not built into kernel??

**Comment by pprindeville** [comment]
> starting rngd before kmods are loaded prevents the use of any rng that requires a kmod to be loaded before its device appears, e.g. ttyACM0.
> 
> moving it to START=11 allows it to be used with devices like this. would that be a good compromise between starting early enough and remaining compatible for hwrng drivers not built into kernel?

Do serial-to-USB devices generate enough events quickly enough to be meaningful sources of entropy?

**Comment by tofurky** [comment]
https://www.gniibe.org/memo/development/gnuk/rng/neug.html
https://ubld.it/products/truerng-hardware-random-number-generator/

etc.

**Comment by pprindeville** [comment]
> https://www.gniibe.org/memo/development/gnuk/rng/neug.html
> https://ubld.it/products/truerng-hardware-random-number-generator/
> 
> etc.

Hmm.

> "The TrueRNG is a low-cost USB Hardware Random Number Generator that provides more than 350 kilobits/second of true random numbers."

**Pull request merged**